stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,1805 @@
+        let Some(origin_signer_id) = self.signer_address_ids.get(&origin_address) else {
+            debug!(
+                "Signer #{}: Unrecognized origin address ({origin_address}). Filtering ({}).",
+                self.signer_id,
+                transaction.txid()
+            );
+            return None;
+        };
+        let Ok(account_nonce) = retry_with_exponential_backoff(|| {
+            stacks_client
+                .get_account_nonce(&origin_address)
+                .map_err(backoff::Error::transient)
+        }) else {
+            warn!(
+                "Signer #{}: Unable to get account for transaction origin address: {origin_address}. Filtering ({}).",
+                self.signer_id,
+                transaction.txid()
+            );
+            return None;
+        };
+        if origin_nonce < account_nonce {
+            debug!("Signer #{}: Received a transaction with an outdated nonce ({account_nonce} < {origin_nonce}). Filtering ({}).", self.signer_id, transaction.txid());
+            return None;
+        }
+        let Ok(valid) = retry_with_exponential_backoff(|| {
+            self.verify_payload(stacks_client, &transaction, *origin_signer_id)
+                .map_err(backoff::Error::transient)
+        }) else {
+            warn!(
+                "Signer #{}: Unable to validate transaction payload. Filtering ({}).",
+                self.signer_id,
+                transaction.txid()
+            );
+            return None;
+        };
+        if !valid {
+            debug!(
+                "Signer #{}: Received a transaction with an invalid payload. Filtering ({}).",
+                self.signer_id,
+                transaction.txid()
+            );
+            return None;
+        }
+        debug!(
+            "Signer #{}: Expect transaction {} ({transaction:?})",
+            self.signer_id,
+            transaction.txid()
+        );
+        Some(transaction)
+    }
+
+    ///Helper function to verify the payload contents of a transaction are as expected
+    fn verify_payload(
+        &self,
+        stacks_client: &StacksClient,
+        transaction: &StacksTransaction,
+        origin_signer_id: u32,
+    ) -> Result<bool, ClientError> {
+        let TransactionPayload::ContractCall(payload) = &transaction.payload else {
+            // Not a contract call so not a special cased vote for aggregate public key transaction
+            return Ok(false);
+        };
+
+        if payload.contract_identifier() != boot_code_id(SIGNERS_VOTING_NAME, self.is_mainnet)
+            || payload.function_name != VOTE_FUNCTION_NAME.into()
+        {
+            // This is not a special cased transaction.
+            return Ok(false);
+        }
+        let Some((index, _point, round, reward_cycle)) =
+            Self::parse_function_args(&payload.function_args)
+        else {
+            // The transactions arguments are invalid
+            return Ok(false);
```

Comment:
This error definitely can happen. A malicious or malformed signer could easily create an vote aggregate transaction with invalid args in an attempt to clog the block or because they just had some sort of error that caused them to create invalid transaction args.

Reply:
Does the Clarity type checker validate these transactions at all before they get here? If not, that might be something we want to happen, but it doesn't have to be in here. If the type checker isn't checking this, then yeah there can be anything in here.

If/when the type checker checks this, then we can still have:

- signer index, round, or reward cycle > u64::max
- aggregate_public_key can't be deserialized into `Point`

So yeah you're right that it's definitely best to gracefully exit, regardless of the type checker

Reply:
These transactions are ones that haven't been executed yet because we have not actually included them in a block yet. There is no way to verify they are correct until you actually execute them I do believe which would clog the blog with invalid transactions. The clarity end will deny them as they should. but this is simply a check to prevent malicious signers from spamming a block with bad transactions